keps/sig-cluster-lifecycle/kubeadm/4471-cp-join-kubelet-local-apiserver/README.md

```diff
@@ -0,0 +1,665 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+As explained above two minor changes are required to implement the required changes.
+
+**1. Make the kubelet point to the local kube-apiserver**
+
+To make the kubelet point to the local apiserver, the file for kubelet's bootstrap kubeconfig
+needs to get adjusted, which gets created by kubeadm during the
+`KubeletStartJoinPhase` ([xref](https://github.com/kubernetes/kubernetes/blob/caf5311/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L122-L125)).
+
+This creates the following chicken-egg issue:
+
+- After starting the kubelet, kubeadm waits in the `KubeletStartPhase` for it to bootstrap itself, before it would continue to start etcd in the `ControlPlaneJoinPhase`.
+- Due to the above change, the kubelet will never be able to bootstrap itself while pointing at the local kube-apiserver, because the local kube-apiserver will never get ready unless etcd is available locally.
+
+The second change address this issue.
+
+**2. Introducing and reordering phases for kubeadm join**
+
+To fix the above chicken-egg issue some operations of kubeadm need to get rearranged during `kubeadm join`.
+
+To be more precise: parts of the `KubeletStartPhase` need to run after the `EtcdLocalSubPhase` (which is part of the `ControlPlaneJoinPhase`).
+
+The current phases of `kubeadm join` are the following:
+
+1. `PreflightPhase`
+2. `ControlPlanePreparePhase`
+3. `CheckEtcdPhase`
+4. `KubeletStartPhase`
+5. `ControlPlaneJoinPhase`
+
+First the existing `KubeletStartPhase` gets split up and the code which waits
+for the kubelet's bootstrap to complete into a separate phase named `KubeletWaitBootstrapPhase`
+gets extracted.
+
+1. `PreflightPhase`
+2. `ControlPlanePreparePhase`
+3. `CheckEtcdPhase`
+4. `KubeletStartPhase`
+5. **`KubeletWaitBootstrapPhase`**
+6. `ControlPlaneJoinPhase`
+
+Second the `EtcdLocalSubphase` which is the first part of the `ControlPlanJoinPhase` gets moved
+to a new phase `ControlPlaneJoinEtcdPhase` and added between the `KubeletStartPhase` and
+`KubeletWaitBootstrapPhase`.
+
+1. `PreflightPhase`
+2. `ControlPlanePreparePhase`
+3. `CheckEtcdPhase`
+4. `KubeletStartPhase`
+5. **`ControlPlaneJoinEtcdPhase`**
+6. `KubeletWaitBootstrapPhase`
+7. `ControlPlaneJoinPhase`
+
+Actions to preserve the old behavior when the feature gate is disabled:
+
+- There should be no duplication of code, instead the relevant shared code parts should be moved
+  to private functions so both phases could execute them, depending on the feature gate.
+- The new phases very lightweight and make use of the private functions.
+- The phases `KubeletStartPhase` and `ControlPlaneJoinPhase` should behave as before and run the above introduced private functions.
+- Nothing should be done during the new phases `ControlPlaneJoinEtcdPhase` and `KubeletWaitBootstrapPhase`.
```

Comment:
one thing we can do is to add "EXPERIMENTAL: some phase description" in the descriptions for all new phases until the FG graduates to GA.